utcp.c

```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
 		case FIN_WAIT_2:
 			// We shouldn't need to retransmit anything in this state.
 #ifdef UTCP_DEBUG
+			debug("Error: retransmit unexpected connection state %p %s\n", c, strstate[c->state]);
 			abort();
 #endif
-			timerclear(&c->rtrx_timeout);
-			break;
+			stop_retransmit_timer(c);
+			goto cleanup;
 	}
 
+	start_retransmit_timer(c);
+	utcp->rto *= 2;
+	if(utcp->rto > MAX_RTO)
+		utcp->rto = MAX_RTO;
+	c->rtt_start.tv_sec = 0; // invalidate RTT timer
+
+cleanup:
 	free(pkt);
 }
 
+/* Update receive buffer and SACK entries after consuming data.
+ *
+ * Situation:
+ *
+ * |.....0000..1111111111.....22222......3333|
+ * |---------------^
+ *
+ * 0..3 represent the SACK entries. The ^ indicates up to which point we want
+ * to remove data from the receive buffer. The idea is to substract "len"
+ * from the offset of all the SACK entries, and then remove/cut down entries
+ * that are shifted to before the start of the receive buffer.
+ *
+ * There are three cases:
+ * - the SACK entry is ahead of ^, in that case just change the offset.
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
```

Comment:
On Thu, Oct 29, 2015 at 05:56:09AM -0700, vollkorn1982 wrote:

> > + * |.....0000..1111111111.....22222......3333|
> > + * |---------------^
> > + *
> > + * There are three cases:
> > + * - the SACK entry is ahead of ^, in that case just change the offset.
> > + * - the SACK entry starts before and ends after ^, so we have to
> > + *   change both its offset and size.
> > + * - the SACK entry is completely before ^, in that case delete it.
> 
> I'm confused by `ahead` and `before` in this context. I'd think both would mean the same direction in the buffer, i.e. left of ^.

s/ahead/after/, and after being right of ^.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>